### PR TITLE
EES-5017, EES-5016 Add data set GET query endpoint

### PR DESCRIPTION
**Pull request by ntsim**
This PR adds the data set GET query endpoint - `/api/v1/data-sets/{dataSetId}/query`. This provides functionality to let users make data set queries using query parameters e.g. `?indicators=headcount&filters.in=kaNhs,zvUFQ&sorts=geographicLevel|Asc`

Note that this doesn't provide test coverage for most of this functionality. This will be covered in EES-5015.

## Query parsing

Parsing data set queries to this endpoint roughly proceeds like so:

1. Validate the GET query string (as `DataSetGetQueryRequest`)
2. Convert to `DataSetQueryRequest` (will be used in POST query variant) in `DataSetQueryService`
3. Parse and validate query criteria using `DataSetQueryParser`, producing a SQL WHERE fragment
4. Parse and validate indicators from data set query
5. Parse and validate sorts and their respective column orderings from the data set query
6. Execute data set query using all requisite parts (where, columns and sorts)
7. Return the paginated results

We've now implemented the majority of the base query criteria parsing functionality to handle the various facet types and their comparators. This doesn't include handling of more complex operators like `and`, `or` and `not`. This will be implemented in EES-4722.

## Dapper and SQL builders

As there is no ORM support for DuckDB, the querying functionality has been implemented using a combination of [Dapper](https://github.com/DapperLib/Dapper) and [InterpolatedSql](https://github.com/Drizin/InterpolatedSql).

Dapper itself is very simplistic and only works with raw SQL strings out of the box. This isn't particularly safe in the context of SQL injection, so we've complemented it with InterpolatedSql which provides protection using its `SqlBuilder` API e.g.

```cs
var myParam = "test";
var builder = new SqlBuilder($"SELECT * FROM something WHERE field = {myParam}");

var sql = builder.Build(); 
sql.Sql; // Outputs: "SELECT * FROM something WHERE field = @p0"
sql.SqlParameters; // Outputs: Single parameter with value "test"
```

For real usages of this, check out the various repositories that have been created. We use a modified variant of InterpolatedSql's Dapper integration to make it a bit nicer to build and execute queries.

## DuckDB issues

Unfortunately, there were some issues with DuckDB that we've had to work around, primarily around named parameters. Dapper only really works with named parameters, but this seems to cause issues with the DuckDB.NET library that we're using. 

When using named parameters, DuckDB.NET would consistently fail to throw exceptions around parameter binding (need to create an issue for this), or [memory access violations when debugging](https://github.com/Giorgi/DuckDB.NET/issues/178). 

I'm communicating with the library maintainer to see if there's anything we can do here. For the time-being, I've been able to implement workarounds by wrapping the DuckDB.NET database connection code and forcing the use of positional auto-incrementing parameters (i.e. `?` parameters) in generated SQL.

## MiniProfiler

To help with debugging and performance, we've added [MiniProfiler](https://miniprofiler.com/) which lets us profile our code. This is particularly useful for data set queries, where it can help identify performance bottlenecks.

MiniProfiler is enabled via the `MiniProfiler.Enabled` app setting, so it can be disabled or enabled on a per-environment basis.

## Related changes

- Fixes `GeographicLevels` in the data set meta summary not being populated correctly from a data set version. This now uses the `GeographicLevelMeta` instead of `LocationMetas` to populate it.

## Other changes

- Improve error details for location, time period and sort string validators by now showing the parsed fields.
- Fixed null pointer exception being thrown when gathering message placeholders from `ValidationFailure` in `FluentValidationActionFilter`

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+            {
+                if (row[locationColumn] is int locationId)
+                {
+                    locationOptionIds.Add(locationId);
+                }
+            }
+
+            foreach (var filterColumn in columnsByType.Filters)
+            {
+                if (row[filterColumn] is int locationId)
```
Comment (jack-hive) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs:340]: `filterId`

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+        using var _ = MiniProfiler.Current.Step(
+            $"{nameof(DataSetQueryService)}.{nameof(MapQueryResults)}");
+
+        var locationColumnsByCode = columnsByType.LocationLevels
+            .ToDictionary(
+                level => level.GetEnumValue(),
+                DataTable.Cols.LocationId
+            );
+
+        var rowList = new List<IDictionary<string, object?>>();
```
Comment (jack-hive) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs:321]: Can we remove `rowList` and just use `rows` on line 362?

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Requests/DataSetQuerySort.cs [hunk 3]
```diff
@@ -35,12 +35,12 @@ public record DataSetQuerySort
...
     {
-        var parts = sort.Split('|');
+        var directionDelimiter = sort.LastIndexOf('|');
 
         return new DataSetQuerySort
         {
-            Field = parts[0],
-            Direction = parts[1]
+            Field = sort[..directionDelimiter],
+            Direction = sort[(directionDelimiter + 1)..]
```
Comment (jack-hive) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Requests/DataSetQuerySort.cs:43]: Was this change necessary or did you just think it was nicer? Doesn't look necessary to me but I don't mind either way

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Scripts/Commands/SeedDataCommand.cs [hunk 4]
```diff
@@ -855,12 +855,40 @@ private async Task OutputParquetFiles()
...
+            {
+                var insertRow = appender.CreateRow();
+
+                insertRow.AppendValue(meta.PublicId);
+                insertRow.AppendValue(meta.Label);
+                insertRow.AppendValue(meta.Unit?.GetEnumLabel() ?? string.Empty);
+                insertRow.AppendValue(meta.DecimalPlaces);
+                insertRow.EndRow();
+            }
+        }
```
Comment (jack-hive) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Scripts/Commands/SeedDataCommand.cs:889]: am I right in saying we currently only use this to get a list of all the indicator column names (i.e. `PublicIds`) in the data set when executing the query logic?

And we don't actually use the `Label`, `Unit`, or `DecimalPlaces` yet?

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/ViewModels/ErrorViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Api.ViewModels;
+
+public record ErrorViewModel : Common.ViewModels.ErrorViewModel
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/ViewModels/ErrorViewModel.cs]: There's some weaknesses with the common `ErrorViewModel` in that its fields are essentially all optional.

This is problematic as it means as this means that the view model is potentially going to be wrong in the generated docs.

There's also some other issues with the current generated docs in terms of required properties. Instead of doing this here, I'm going to try and tackle this in this subtask: https://dfedigital.atlassian.net/browse/EES-5057

File: src/GovUk.Education.ExploreEducationStatistics.Common/ViewModels/ErrorViewModel.cs [hunk 6]
```diff
@@ -2,12 +2,15 @@
...
+/// <summary>
+/// Describes an error that occurred and needs attention to rectify.
+/// </summary>
 public record ErrorViewModel
 {
     /// <summary>
     /// The error message.
     /// </summary>
-    public string Message { get; init; } = string.Empty;
+    public virtual string Message { get; init; } = string.Empty;
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Common/ViewModels/ErrorViewModel.cs]: As per other comment - this is due to issues with the current view model. We ideally want to be able to extend this to create a more correct view model for the public API.

For now, I'm leaving this to implement it in https://dfedigital.atlassian.net/browse/EES-5057 instead.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Validators/LocationStringValidators.cs [hunk 7]
```diff
@@ -148,32 +151,28 @@ private static int GetMaxLength(ParsedLocation location)
...
-        : InvalidErrorDetail<string>(Value);
+    public record AllowedPropertyErrorDetail(ParsedLocation Value, IEnumerable<string> AllowedProperties)
+        : InvalidErrorDetail<ParsedLocation>(Value);
 
-    public record MaxValueLengthErrorDetail(string Value, int MaxLength) : InvalidErrorDetail<string>(Value);
+    public record MaxValueLengthErrorDetail(ParsedLocation Value, int MaxValueLength)
+        : InvalidErrorDetail<ParsedLocation>(Value);
 
-    private record ParsedLocation
+    public record ParsedLocation
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Validators/LocationStringValidators.cs:163]: Yep it needs to be public it's being used by the error detail types above which are themselves public.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/ViewModels/TimePeriodViewModel.cs [hunk 8]
```diff
@@ -7,6 +9,7 @@ namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Api.ViewModels;
 /// </summary>
 public record TimePeriodViewModel
 {
+    [JsonConverter(typeof(EnumToEnumValueJsonConverter<TimeIdentifier>))]
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/ViewModels/TimePeriodViewModel.cs:12]: As other comment - this is for JSON deserialisation, not Swagger docs

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+        if (!VersionUtils.TryParse(dataSetVersion, out var version))
+        {
+            return new NotFoundResult();
+        }
+
+        return await publicDataDbContext.DataSetVersions
+            .SingleOrNotFoundAsync(
+                dsv => dsv.Id == dataSetId
+                       && dsv.VersionMajor == version.Major
+                       && dsv.VersionMinor == version.Minor,
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs]: Don't think this was the wrong amount of indents, but not relevant anymore as I've changed it to separate `Where` calls as per your suggestion in another comment anyway.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Model/DuckDb/DuckDbSqlBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+public class DuckDbSqlBuilder : InterpolatedSqlBuilderBase<DuckDbSqlBuilder, IInterpolatedSql>
+{
+    private static readonly InterpolatedSqlBuilderOptions DefaultOptions = new()
+    {
+        // Use auto-incrementing positional parameters (i.e. ?), to avoid issues
+        // with named parameters which don't seem to work reliably in all environments.
+        // See: https://github.com/Giorgi/DuckDB.NET/issues/178
+        DatabaseParameterSymbol = "?",
+        CalculateAutoParameterName = (_, _) => string.Empty,
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Model/DuckDb/DuckDbSqlBuilder.cs:16]: Yes, otherwise we end up with placeholders like `?p0`, `?p1`, which aren't valid

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Model/DuckDb/DuckDbSqlBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                index += 1;
+            }
+        }
+        else
+        {
+            base.AppendArgument(argument, alignment, format);
+        }
+    }
+
+    public override IInterpolatedSql Build() => base.AsSql();
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Model/DuckDb/DuckDbSqlBuilder.cs:72]: We can't call `base.Build` because it's an abstract method and we have to implement it. In this case, we implement it by calling `base.AsSql`, which is essentially what the official SqlBuilder for Dapper does.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        HashSet<string> filterOptionIds,
+        string path,
+        bool negate = false)
+    {
+        var builder = new DuckDbSqlBuilder();
+
+        var options = filterOptionIds
+            .Distinct()
+            .Select(id => _filterOptionsByPublicId.GetValueOrDefault(id))
+            .OfType<ParquetFilterOption>()
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs:107]: If we did this, it'd be likely for exceptions to thrown as the `id` may not actually exist as a key of `_filterOptionsByPublicId`.

We do it the current way so that we get nulls if the key doesn't exist, and then we filter out the nulls using `OfType`.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+            .ToList();
+
+        if (options.Count < filterOptionIds.Count)
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs:110]: `filterOptionIds` is already a set, so all the values are already distinct :wink:

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        var fragments = new List<IInterpolatedSql>();
+
+        var parsedEq = facets.GeographicLevels?.ParsedEq;
+
+        if (parsedEq is not null)
+        {
+            fragments.Add(
+                EqFragment(
+                    geographicLevel: parsedEq.Value,
+                    path: "geographicLevels.eq"
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs:29]: Yep, we need this as it concatenates path segments together. We need the `path` as this is provided with any errors or warnings and directs the user to the part of the request that was causing the problem.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            _queryState.Warnings.Add(CreateNotFoundWarning(filterOptionIds, path));
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs:112]: As per other comment - it's already a `HashSet` so no need for `Distinct`

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/LocationFacetsParser.cs [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model.Parquet;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model.Parquet.Tables;
+using InterpolatedSql;
+
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Api.Services.Query;
+
+internal class LocationFacetsParser : IFacetsParser
+{
+    private readonly QueryState _queryState;
+    private readonly Dictionary<GeographicLevel, List<ParquetLocationOption>> _locationOptionMetas;
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/LocationFacetsParser.cs:20]: I've gone with `_allowedLocationOptionsByLevel`

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        if (levels.Count == 0)
+        {
+            return builder
+                .AppendLiteral(negate ? "true" : "false")
+                .Build();
+        }
+
+        var parameters = levels
+            .Select(level => level.GetEnumLabel())
+            .Cast<object>()
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs:124]: Nope we don't, looks like a hangover from a previous iteration of this!

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+        string path,
+        bool negate = false)
+    {
+        var builder = new DuckDbSqlBuilder();
+
+        var ids = timePeriods
+            .Select(timePeriod => new TimePeriodKey(timePeriod))
+            .Where(_timePeriodMetas.ContainsKey)
+            .Select(timePeriod => _timePeriodMetas[timePeriod].Id)
+            .ToList();
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs]: This isn't necessary as it's calculated from `timePeriods` which is a `HashSet` and `_allowedTimePeriods` (renamed as per other comment) is a dictionary which shouldn't have duplicates either.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+        if (ids.Count < timePeriods.Count)
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs:178]: As above - doesn't seem necessary

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+        {
+            _queryState.Warnings.Add(CreateNotFoundWarning(timePeriods, path));
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs:180]: As above - doesn't seem necessary

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+        if (!_timePeriodMetas.TryGetValue(new TimePeriodKey(timePeriod), out var meta))
+        {
+            _queryState.Warnings.Add(CreateNotFoundWarning([timePeriod], path));
+
+            return builder
+                .AppendLiteral("false")
+                .Build();
+        }
+
+        builder += $"{DataTable.Ref().TimePeriodId:raw} {comparator:raw} {meta.Id}";
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs:160]: Yep, exactly :smile:

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+        bool negate = false)
+    {
+        var builder = new DuckDbSqlBuilder();
+
+        if (!_filterOptionsByPublicId.TryGetValue(filterOptionId, out var option))
+        {
+            _queryState.Warnings.Add(CreateNotFoundWarning([filterOptionId], path));
+
+            return builder
+                .AppendLiteral(negate ? "true" : "false")
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs:88]: As discussed in Slack last week, this is correct behaviour as we're concatenating facet fragments together with `AND`. If we're looking for something that doesn't exist, the criteria should resolve to false overall.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+        return (await command.QueryAsync(cancellationToken: cancellationToken))
+            .Cast<IDictionary<string, object?>>();
+    }
+
+    public async Task<ISet<string>> ListColumns(
+        DataSetVersion dataSetVersion,
+        CancellationToken cancellationToken = default)
+    {
+        var command = duckDbConnection.SqlBuilder(
+            $"DESCRIBE SELECT * FROM '{parquetPathResolver.DataPath(dataSetVersion):raw}' LIMIT 1");
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs:105]: `LIMIT 1` just means that only one row is returned, rather than all rows. We still get a complete set of column information.

I'm not entirely sure if adding the `LIMIT 1` really makes a difference as we're using `DESCRIBE`. I'm not sure what happens under the hood with DuckDB, but I'd hope it doesn't actually try and retrieve all rows when using `DESCRIBE SELECT *`, but in-case it does, I've added in the `LIMIT 1` as a precaution.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryParser.cs [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        return facets;
+    }
+
+    private IInterpolatedSql ParseCriteriaFragment(
+        DataSetQueryCriteria criteria,
+        IEnumerable<IFacetsParser> facetParsers)
+    {
+        var path = "";
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryParser.cs:68]: Yep, forward planning as users will be able to create more complex queries with the POST endpoint so the path will be able to more precisely say where any problems are occurring in the query.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs (hunk 2 above)
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetQueryService.cs:371]: This was to avoid multiple enumeration warnings in Rider as we're working with `IEnumerable` and I didn't want to have allocate 

However, I've now changed the repository methods to return `IList` instead of `IEnumerable` so we don't need this any more.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Requests/DataSetQuerySort.cs (hunk 3 above)
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Requests/DataSetQuerySort.cs:45]: Yep this is necessary as I've changed up sort fields for locations. This is now `location|{level}|{direction}` e.g. `location|LA|Desc`, `location|REG|Desc`, etc.

This means there can now be either 2 or 3 `|` characters in the sort string.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Validators/LocationStringValidators.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Validators/LocationStringValidators.cs:163]: ah yes

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs:107]: Sorry, I misinterpreted the syntax. I thought it was going to describe all the columns (each of which would be a row of descriptive data) - and then take only one of those resulting rows. Which would end up just being the description for 1 column and ignoring the others!

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/FilterFacetsParser.cs:109]: facepalm moment there.....

I totally missed that!

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs:178]: yep, my bad

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/ViewModels/DataSetQueryResultViewModels.cs [hunk 25]
```diff
@@ -29,6 +31,7 @@ public record DataSetQueryResultViewModel
     /// <summary>
     /// The geographic level relevant to this result's data.
     /// </summary>
+    [JsonConverter(typeof(EnumToEnumValueJsonConverter<GeographicLevel>))]
```
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/ViewModels/DataSetQueryResultViewModels.cs:31]: I was definitely a bit tired when reviewing these bits of the code.....

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Model/DuckDb/DuckDbSqlBuilder.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Model/DuckDb/DuckDbSqlBuilder.cs:72]: of course

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs:29]: Sorry, I think you misinterpreted what I was asking.

I was suggesting changing this line to `path: QueryUtils.Path(path, "geographicLevels.eq")` rather than the current `path: "geographicLevels.eq"`

[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs:29]: currently the `path` parameter passed into the `Parse` method isn't being used

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/TimePeriodFacetsParser.cs]: yep, this was me being tunnel-visioned on all the `HashSet` stuff. For some reason I just wasn't seeing them!

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/Query/GeographicLevelFacetsParser.cs:29]: Ah sorry my bad, didn't read your comment correctly! I've added this in now, nice catch!

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+        var command = duckDbConnection.SqlBuilder(
+            $"""
+             SELECT DISTINCT {IndicatorsTable.Cols.Id:raw}
+             FROM '{parquetPathResolver.IndicatorsPath(dataSetVersion):raw}'
+             """);
+
+        var indicators = await command.QueryAsync<string>(cancellationToken: cancellationToken);
+
+        return indicators.ToHashSet();
+    }
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetDataRepository.cs]: I did originally have something like this, but was a bit on the fence about it and eventually settled on having them in `ParquetDataRepository`.

Given your suggestion, I've refactored them into their respective repositories with the suggested renames. Note that some of the existing methods needed to change too e.g. `List` to `ListOptions`.

File: src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetFilterRepository.cs [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+        return (await command.QueryAsync<T>(cancellationToken: cancellationToken)).AsList();
+    }
+
+    public async Task<ISet<string>> ListFilterIds(
+        DataSetVersion dataSetVersion,
+        CancellationToken cancellationToken = default)
+    {
+        var command = duckDbConnection.SqlBuilder(
+            $"""
+             SELECT DISTINCT {FiltersTable.Cols.ColumnName:raw}
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetFilterRepository.cs:103]: @jack-hive penny for your thoughts as I've gone back and forth on this in my head a few times, but what do you think about this column name?

I'm kinda considering renaming it to  `FilterId` (`filter_id` in the table itself). However, this is a touch confusing given the table itself is called `filters.parquet` and we have other id columns too.

Like I said, a bit on the fence about this. It might be fine as-is, but maybe not. Any thoughts?

Comment (jack-hive) [reply src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Repository/ParquetFilterRepository.cs:103]: Hmm......

I'm a bit on the fence too.

I think `FilterId` makes more sense in terms of the code - but is less intuitive in terms of the domain context.

Maybe `CsvColumnName` might be more explicit than `ColumnName` and intuitive.

I think I'd either choose `CsvColumnName` or `FilterId`

**Review by jack-hive — APPROVED** [review]

**Review by jack-hive — APPROVED** [review]
